lib/python/rose/apps/rose_bunch.py

```diff
@@ -0,0 +1,458 @@
+                              name TEXT,
+                              status TEXT,
+                              PRIMARY KEY(name))""")
+
+        if self.T_CONFIG not in existing:
+            self.conn.execute("""CREATE TABLE """ + self.T_CONFIG + """ (
+                              key TEXT,
+                              value TEXT,
+                              PRIMARY KEY(key))""")
+        self.conn.commit()
+        return
+
+    def add_command(self, name):
+        """Add a command to the commands table"""
+        i_stmt = "INSERT OR REPLACE INTO " + self.T_COMMANDS + " VALUES (?, ?)"
+        self.conn.execute(i_stmt, [name, self.S_STARTED])
+        self.conn.commit()
+        return
+
+    def clear_command_states(self):
+        """Deletes all recorded command entries"""
+        d_stmt = "DELETE FROM " + self.T_COMMANDS
+        self.conn.execute(d_stmt)
+        self.conn.commit()
+        return
+
+    def close(self):
+        """Close database connection"""
+        if self.conn is not None:
+            self.conn.close()
+            self.conn = None
+
+    def update_command_state(self, name, state):
+        """Update command state in CMDS table"""
+        u_stmt = "UPDATE " + self.T_COMMANDS + " SET status=? WHERE name==?"
+        self.conn.execute(u_stmt, [state, name])
+        self.conn.commit()
+        return
+
+    def check_has_succeeded(self, name):
+        """See if a named command reached the "pass" state"""
+        s_stmt = ("SELECT * FROM " + self.T_COMMANDS +
+                  " WHERE name==? AND status==?")
+        s_stmt_args = [name, self.S_PASS]
+        has_succeeded = False
+        for s_row in self.conn.execute(s_stmt, s_stmt_args):
+            has_succeeded = True
+        return has_succeeded
+
+    def flatten_config(self, config):
+        """Flatten down a config node to set of keys and values for recording"""
+        flat = {}
+        keys_and_nodes = list(config.node.walk())
+        for keys, node in keys_and_nodes:
+            if not isinstance(node.value, dict):
+                if node.state not in [rose.config.ConfigNode.STATE_USER_IGNORED,
+                                      rose.config.ConfigNode.STATE_SYST_IGNORED]:
+                    flat["_".join(keys)] = node.value
+        return flat
+
+    def record_config(self, config, clear_db=False):
+        """Take in a conf_tree object and record the entries"""
+
+        if clear_db:
+            d_stmt = "DELETE FROM " + self.T_CONFIG
+            self.conn.execute(d_stmt)
+            self.conn.commit()
+
+        res = self.flatten_config(config)
+
+        args = []
+
+        for key, value in res.items():
+            args.append((key, value))
+
+        i_stmt = "INSERT OR REPLACE INTO " + self.T_CONFIG + " VALUES (?, ?)"
+        self.conn.executemany(i_stmt, args)
+        self.conn.commit()
+        return
+
+    def same_last_config(self, current):
```

Comment:
`cmp_prev_config`? (The word `previous` is normally clearer in meaning than `last`.)